Trigger full sweep when testing garbage collection

jonathanhefner opened the pull request:
This ensures that the checker will be collected if it can be collected.

This addresses intermittent CI failures such as: https://buildkite.com/rails/rails/builds/79596#0051e91f-12bd-420f-979f-dd9c9ddb6f5b/1076-1085

---

I cannot reproduce the failure locally, so I am unsure whether this will completely fix it, but it is worth trying.

jonathanhefner committed:
Trigger full sweep when testing garbage collection

This ensures that the checker will be collected if it can be collected.

This addresses intermittent CI failures such as:
https://buildkite.com/rails/rails/builds/79596#0051e91f-12bd-420f-979f-dd9c9ddb6f5b/1076-1085

jonathanhefner commented:
Apparently insufficient: https://buildkite.com/rails/rails/builds/79633#29d8170e-7667-4334-a30b-72e2db980867/1096-1106 :disappointed:

p8 commented:
I can reproduce the error by adding the following line to the beginning of the "can be garbage collected" test.

```ruby
    Thread.current.thread_variable_set(:current_attributes_instances, { })`
``

jonathanhefner committed:
Isolate garbage collection test in new thread

jonathanhefner commented:
> I can reproduce the error by adding the following line to the beginning of the "can be garbage collected" test.
> 
> ```ruby
>     Thread.current.thread_variable_set(:current_attributes_instances, { })
> ```

Very interesting!!  :eyes:   I tried that, but still cannot. :cry:  What version of Ruby are you running?  I tried with 2.7.2 and 3.0.1.  I have not tried with 2.7.4 or 3.0.2, which were released a few weeks ago.

p8 commented:
ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-darwin20]

p8 commented:
I can reproduce it on the following as well:
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin20]

byroot commented:
One trick to debug these kind of issues.

```ruby
require 'objspace'
puts ObjectSpace.dump(object_i_want_gced)
ObjectSpace.dump_all(output: "/tmp/heap.dump")
```

Then use https://github.com/csfrancis/harb, to get the "rootpath" of the object that doesn't want to die.

p8 commented:
@jonathanhefner I'll have a try at that.

p8 commented:
@jonathanhefner With c2ed965e59da4f1e738de56342b643f545e125bc the assertion still fails, but in the Thread.
So the test is green.

Clearing the thread_variable at the top of the test fixes it for me:
``` 
Thread.current.thread_variable_set(:current_attributes_instances, nil)
```

byroot commented:
> Clearing the thread_variable at the top of the test fixes it for me:

We must be missing something. Can you print that thread variable before clearing it? We must be leaking state in the thread variables.

p8 commented:
@byroot In this case I'm setting it myself in the line above it.
I'm currently running a build that hopefully outputs the contents of the thread_variable.

jonathanhefner committed:
Move assertion out of thread

jonathanhefner commented:
> With [c2ed965](https://github.com/rails/rails/commit/c2ed965e59da4f1e738de56342b643f545e125bc) the assertion still fails, but in the Thread.
> So the test is green, even though the assertion fails.

Do you mean the `assert_not_empty listener_threads` assertion inside `Thread.new` fails?  That is a new assertion I just added.  I should have added it outside of the thread block, but, regardless, I'm not sure why it would be failing.  However, if the `assert_not checker_ref.weakref_alive?` is passing, that is progress.

I've moved all the assertions into the main thread in https://github.com/rails/rails/pull/42877/commits/fbffa1062386a4f654430b420fb3771e9776e041.  I've also created a *non-weak* ref to the checker inside the thread, in case that was somehow related to `assert_not_empty listener_threads` failing.

I'll continue to dig...

p8 commented:
Ok, some success 🎉 
The thread_variables aren't properly cleared:
https://buildkite.com/rails/rails/builds/79641#67006cbd-d1ea-491b-b78e-2fad59717ec3/994-1001

p8 commented:
> Do you mean the assert_not_empty listener_threads assertion inside Thread.new fails? 
Yes, that assertion failed.

p8 commented:
Here is a run that failed and has the thread variable set. The debug output was in an `ensure` block
https://buildkite.com/rails/rails/builds/79664#620bcdde-f4e7-4433-9af4-6186b49b9e18/1076-1084

p8 commented:
If I assert the Session.current is empty when using thread_variables, the tests will fail:
https://github.com/rails/rails/pull/42878/files#diff-126362eb3c6e6f212a06775703bf4b03ca9bf7b83b06661c49672c55dfba5362
https://buildkite.com/rails/rails/builds/79666#cc9ada15-ca01-43ec-9c54-699c4480d51e/1076-1084
I'm not sure if this is expected behaviour and we should just clear it in the test, or if there are underlying problems.
cc @etiennebarrie

etiennebarrie commented:
Oh right sorry, that test is definitely leaking state. I'll review https://github.com/rails/rails/pull/42880.
